pppd/sys-linux.c

```diff
@@ -140,6 +140,9 @@
 #define NETLINK_CAP_ACK 10
 #endif
 #endif
+#ifdef RTM_GETSTATS
```

Comment:
Now I checked that support for `RTM_GETSTATS` was added in v4.7-rc1 (commit 10c9ead9f3c6). And support for older kernels is required. There are still officially supported v4.4 and plus enterprise distros have even older supported kernel versions.

Reply:
And support for `IFLA_STATS64` and `struct rtnl_link_stats64` was added in v2.6.35-rc1 (commit 10708f37ae72).

Reply:
Wish I had older kernels lying around ... how about this - remove the guarding entirely and if/when someone reports a compile error we figure out those details?

Reply:
Ended up just protecing ```RTM_GETSTATS``` and ```IFLA_STATS_LINK_64```.

I have bumped into a 2.6.32 kernel not too long ago, but I don't see an easy way of protecting IFLA_STATS64 and specifically the rtnl_link.  Best approach there is to probably not compile in the support if kernel headers indicates it's older than 2.6.35.

Reply:
I edited https://github.com/ppp-project/ppp/pull/298#discussion_r765300431 and put there a new version of possible guard. It is untested, but probably something like that could work.

Reply:
Ok, so here is the thing:

```607ca46e97a1b (David Howells                  2012-10-13 10:46:48 +0100  215) struct rtnl_link_stats64 {```

No changes to this structure was made (except to formalize the comments) since.

I don't believe we can request only a portion of the stats, and (NLMSG_OK() would start failing if we only receive a partial packet ... might be worthwhile to actually pad the structure on our end a bit in any case to ensure that we get the whole thing).  Or I need to re-evaluate the use of that MACRO to permit only a portion, but eventually I'm going to have to RX the whole thing anyway since I believe follow-up recv's would fetch the tail of the last kernel-side tx.  Which brings me back to possible need to allocate a slightly larger buffer on the stack so that we don't run into this, although, in such cases I do simply close the socket and fail (which will fall back to sysfs, but if we increase the buffer size we could proceed even if the size of the structure does change for some unforeseen reason).

@pali if you could point me at something that says something different that would be great.

Reply:
So I did this as a test (in a completely different program):

``` 13 struct rtnl_link_stats64_compat {
 14   uint64_t rx_packets;
 15   uint64_t tx_packets;
 16   uint64_t rx_bytes;
 17   uint64_t tx_bytes;
 18 };
 19 #define rtnl_link_stats64 rtnl_link_stats64_compat ```
 
 ```141     struct {
142         struct nlmsghdr nlh;
143         union {
144             struct nlmsgerr err;
145             struct {
146                 struct rtmsg rth;
147                 struct rtnl_link_stats64 stats;
148             };
149         };
150     } resp;```

```202     while (1) {
203         printf("Sending request\n");
204         if (sendmsg(fd, &req_msg, 0) < 0) {
205             perror("sendmsg");
206             return 1;
207         }
208 
209         printf("Receiving message\n");
210         resp_size = recvmsg(fd, &resp_msg, 0);
211         if (resp_size < 0) {
212             perror("recvmsg");
213             return 1;
214         }
215         printf("MSG size: %ld\n", resp_size);
216 
217         if (!NLMSG_OK(&resp.nlh, resp_size)) {
218             printf("NLMSG is NOT OK\n");
219             return 1;
220         }
221 
222         printf("NL sub message: %d\n", resp.nlh.nlmsg_type);
223         switch (resp.nlh.nlmsg_type) {
224         case NLMSG_ERROR:
225             printf("Error code: %d (%s)\n", resp.err.error, strerror(-resp.err.error));
226             goto sleep;
227         case RTM_NEWSTATS:
228             printf("Got stats.\n");
229 #define outp(x) printf("%20s: %5u\n", #x, (unsigned)resp.rth.x)
230             outp(rtm_family);
231             outp(rtm_dst_len);
232             outp(rtm_src_len);
233             outp(rtm_tos);
234             outp(rtm_table);
235             outp(rtm_protocol);
236             outp(rtm_scope);
237             outp(rtm_type);
238             outp(rtm_flags);
239 #undef outp
240 #define outp(x) printf("%20s: %20lu\n", #x, resp.stats.x)
241             outp(rx_packets);
242             outp(tx_packets);
243             outp(rx_bytes);
244             outp(tx_bytes);
245 //          outp(rx_errors);
246 //          outp(tx_errors);
247 //          outp(rx_dropped);
248 //          outp(rx_dropped);
249 #undef outp
250             break;
251         default:
252             printf("Unknown sub type (%d).\n", resp.nlh.nlmsg_type);
253         }
254 
255 sleep:
256         sleep(10);
257     }```

Tried to trim things down as much as possible, I set up a single req_msg for sending every 10 seconds.  Not much in the way of testing here.  NLMSG_OK fails instantly:

```sizeof resp: 64
Sending request
Receiving message
MSG size: 64
NLMSG is NOT OK```

Eliminating the NLMSG_OK and replacing with:

```227         if (resp_size < (int)sizeof(resp.nlh)) {
228             printf("response is smaller than nlh\n");
229             return 1;
230         }```

then just manually checking as we descend that we retrieved at least the bytes required (without using NLMSG_PAYLOAD since that's just retrieving the fields in the nlmsg structure so can't rely on them due to NLMSG_OK being false - we didn't retrieve the whole message), so something like:

```142     struct resp_structure {
143         struct nlmsghdr nlh;
144         union {
145             struct {
146                 struct nlmsgerr err;
147                 char end_err[0];
148             };
149             struct {
150                 struct rtmsg rth;
151                 struct rtnl_link_stats64 stats;
152                 char end_stats[0];
153             };
154         };
155     } resp;
...
242             if (resp_size < (int)offsetof(struct resp_structure, end_stats)) {
243                 printf("Incomplete stats\n");
244                 return 1;
245             }```

Where offsetof is used before accessing the statistics in this case, similar checks for error messages ... so we just can't rely on the MACROs from netlink.h.  Happy to make this adjustments if you prefer.  I think the code just gets nastier and harder to read in order to save a few bytes being copied between kernel/user space ...

Reply:
Perfect testing! This is probably reason why I also have not used `NLMSG_OK` macro...